test/functional/test_framework/key.py

```diff
@@ -444,3 +519,68 @@ def tweak_add(self, tweak):
         ret = ECKey()
         ret.set(tweaked.to_bytes(32, 'big'), self.compressed)
         return ret
+
+def generate_schnorr_nonce():
+    """Generate a random valid Schnorr nonce."""
+    kp = random.randrange(1, SECP256K1_ORDER)
+    assert(kp != 0)
+    R = SECP256K1.affine(SECP256K1.mul([(SECP256K1_G, kp)]))
+    k = kp if jacobi_symbol(R[1], SECP256K1_FIELD_SIZE) == 1 else SECP256K1_ORDER - kp
+    k_key = ECKey()
+    k_key.set(k.to_bytes(32, 'big'), True)
+    return k_key
+
+def aggregate_schnorr_nonces(Rv):
+    """Construct aggregated musig nonce from individually generated nonces."""
+    for idx, R in enumerate(Rv):
```

Comment:
nit: consider using a map-reduce here:

```
from functools import reduce
R_agg = reduce(lambda x, y: SECP256K1.add(x, y), map(lambda x: x.p, Rv))
```

actually, since you've already defined a way to add `ECPubKey`s, you don't even need the map:

```
R_agg = reduce(lambda x, y: x.add(y), Rv).p
```

If you define `__add__` for `ECPubKey`, then this just becomes:

```
R_agg = sum(Rv).p
```

edit: since you're returning a `ECPubKey` object, just set `R_agg` as a `ECPubKey`:

```
R_agg = reduce(lambda x, y: x.add(y), Rv)
```

Reply:
Very nice!